luarules/gadgets/cmd_nano_range_check.lua

```diff
@@ -0,0 +1,156 @@
+function gadget:GetInfo()
+	return {
+		name = "Construction Turrets Range Check",
+		desc = "Stops construction turrets from getting stuck on orders out of reach.",
+		author = "Nehroz",
+		date = "2024.12.01",
+		layer = 0,
+		enabled = true,
+	}
+end
+
+if not gadgetHandler:IsSyncedCode() then return end
+
+local CMD_STOP = CMD.STOP
+local gMaxUnits = Game.maxUnits
+local updateInterval = Game.gameSpeed / 2
+
+local spValidUnitID = Spring.ValidUnitID
+local spValidFeatureID = Spring.ValidFeatureID
+local spGetUnitDefID = Spring.GetUnitDefID
+local spGetUnitSeparation = Spring.GetUnitSeparation
+local spGetUnitFeatureSeparation = Spring.GetUnitFeatureSeparation
+local spGetUnitCommands = Spring.GetUnitCommands
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+
+--- Key is the nano's ID, followed by a set-like with the unitID of target currently inside the nano's range.
+--- Used for CommandArray manipulation
+---@type {[integer] : {integer : {cmdTag : integer, status : integer}}}
+local trackingTable = {}
+local tagUpdateFrameMax = 1 -- Can be bigger than 1 if you want game freezes.
+
+local statuses = {
+	outOfRange = 0,
+	inRange = 1,
+}
+
+local mobileUnits = {}
+local constructionTurretsDefs = {}
+
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	if unitDef.isFactory == false and unitDef.isStaticBuilder then
+		constructionTurretsDefs[unitDefID] = {
+			maxBuildDistance = unitDef.buildDistance + unitDef.radius,
+			buildRange3D = unitDef.buildRange3D,
+		}
+	end
+
+	if unitDef.canMove then
+		mobileUnits[unitDefID] = true
+	end
+end
+
+function gadget:Initialize()
+	gadgetHandler:RegisterAllowCommand(CMD.REPAIR)
+	gadgetHandler:RegisterAllowCommand(CMD.GUARD)
+	gadgetHandler:RegisterAllowCommand(CMD.RECLAIM)
+	gadgetHandler:RegisterAllowCommand(CMD.STOP)
+end
+
+function gadget:GameFrame(frame)
+	local tagUpdates = 0
+	local cmdCache = {}
+
+	for nanoID, targets in pairs(trackingTable) do
+		if nanoID % updateInterval == frame % updateInterval then
+			if next(targets) == nil then
+				trackingTable[nanoID] = nil
+			end
+
+			local nanoDefID = spGetUnitDefID(nanoID)
+			local maxDistance = constructionTurretsDefs[nanoDefID].maxBuildDistance
+
+			for targetID, commandData in pairs(targets) do
+				if spValidUnitID(targetID) then
+					if commandData.cmdTag == -1 then
+						if tagUpdates >= tagUpdateFrameMax then break end
+
+						if not cmdCache[nanoID] then
+							cmdCache[nanoID] = spGetUnitCommands(nanoID, -1)
```

Comment:
Why do you need to get all commands? Why not just look at the current command and drop it if it happens to be out of range. You'd only need to check each order once, using `gadget:UnitCmdDone` to readd a nano to the check list when its current order finishes.

Reply:
Tracking has to occur anyway on a regular basis for the states to update. And the orders are not sorted by in/outside status; possible future optimisation. So just mutating the current command or popping them off would ruin the behaviour and miss units while another unit is indexed before them.
 
That's why a cmd gets removed as they appear, instead of checking over the table on a task completion. This way the tags are pre-cached instead of fetching and creating a whole table each time a task is completed.

When it's possible to get the cmdTags of a certain indexes or last of the current command list, this can be replaced. `_, _, tag = Spring.GetUnitCurrentCommand(unitID, x)` and even be popped off with a UnitCmdDone until the next valid unit is reached. https://github.com/beyond-all-reason/spring/issues/1725
As of now this is the better performance way.

Reply:
Would be great to avoid getting the full queue, but if there's no other way I guess we can live with that for now.